src/templater.rs

```diff
@@ -559,26 +559,31 @@ impl<'a> CommitOrChangeId<'a> {
         hex
     }
 
-    pub fn shortest_prefix_and_brackets(&self) -> String {
+    /// The length of the id printed will be the maximum of `total_len` and the
+    /// length of the shortest unique prefix
+    pub fn shortest_prefix_and_brackets(&self, total_len: i64) -> String {
         let hex = self.hex();
+        let total_len = max(2, total_len) as usize;
         let (prefix, rest) = extract_entire_prefix_and_trimmed_tail(
             &hex,
             self.repo.shortest_unique_id_prefix_len(self.as_bytes()),
-            12 - 2,
+            total_len - 2,
         );
         if rest.is_empty() {
-            prefix.to_string()
+            format!("{prefix:total_len$}")
```

Comment:
I included it explicitly so that it always prints at least `total_len` characters. That made the comment above the function easier to write. This only matters in a very special case where `total_len` is longer than the shortest prefix by 1 or 2 characters, so there's no room for the brackets.

I'm happy to remove the padding. I was going to do it myself, but then I looked at the tests and thought they looked slightly better because of it.

Reply:
FWIW, I'd prefer the length of the bracketed style to not include the brackets. My impression is that people usually choose the hash length based on what's long enough to be unique. Implementation-wise, it also seems better to have the colored and bracketed styles be based on the same data. It would also remove the special case you're talking about, right?

Reply:
Yeah, that's pretty convincing.

Reply:
And the default of `.shortest()` (not the config option) could be `total_len=0` to make it purely the shortest prefix method? Mercurial's default is `minlength=4`, but I feel 0 is better than arbitrary number.

(Not meant to be included in this PR)

Reply:
Done. I'm not sure if one of you is interested in reviewing this after all the changes. If I don't hear from you, I'll merge this in an hour or two.

Reply:
I'm currently looking at #1189, but I'll try to get to this one after that, unless Yuya beats me to it.

Reply:
> And the default of .shortest() (not the config option) could be total_len=0 to make it purely the shortest prefix method?

Sounds good.

Reply:
> I'm currently looking at https://github.com/martinvonz/jj/pull/1189, but I'll try to get to this one after that, unless Yuya beats me to it.

Thanks! I'll wait for you to take a look then, no rush.

Reply:
Looks like Yuya did beat me, so I'll skip. Thanks to both of you!

Reply:
Thanks Yuya!